django/forms/widgets.py

```diff
@@ -762,8 +762,17 @@ class RadioSelect(ChoiceWidget):
     template_name = 'django/forms/widgets/radio.html'
     option_template_name = 'django/forms/widgets/radio_option.html'
 
+    def id_for_label(self, id_, index=None):
+        """
+        Don't include for="field_0" in <label> because clicking such a label
+        would toggle the first checkbox.
```

Comment:
That is a bug in screen readers though, right? Input elements in HTML can have more than on label (see [HTML-AADM 1.0 §5.1.1 (2)](https://www.w3.org/TR/html-aam-1.0/#input-type-text-input-type-password-input-type-search-input-type-tel-input-type-email-input-type-url-and-textarea-element-accessible-name-computation)). But if the reality is that screen readers only use the first label we should deal with that anyway.

Reply:
Good question. It’s also possible for this to be a VoiceOver-only issue – right now it’s very hard for me to test with any other screen reader, so technically I haven’t tested it anywhere else.

Screen readers do have wildly different behaviors for accessible name computation, as well as role, and lots of parts of the ARIA spec, so yes nonetheless it needs dealing with.

I also can’t see `type="radio"` or checkbox in your link, which is a bit puzzling to me.

Reply:
Testing this now – seems like a Safari issue. On macOS:

- In Safari 14.1, the accessible name for the `input` is computed as `Radio choice required:` (field’s label only)
- In Firefox 88, `Radio choice required: One`
- In Chrome 91, `Radio choice required: One`

As far as I know on macOS the only combination of screen reader + browser that’s really prevalent is Safari + VoiceOver, which shows the problem. Nonetheless it’s good to know this might not be as much of an issue for NVDA + Firefox users on Windows, and JAWS + Chrome users.

With this info, we could update the release notes and this comment to state this change addresses a bug in Safari + VoiceOver, and is an improvement for other browser + screen reader combinations. And mention the "clicking on the label" thing as well.

Reply:
I've had a go at rewording the comment here and the release note. Would you like to take another look?

Reply:
Thank you, looks great to me. Sorry it took so long to get to the bottom of this! It’s a good reminder for me to properly caveat my findings for future issues.

Reply:
Here is the relevant WebKit bug: https://bugs.webkit.org/show_bug.cgi?id=152663. Opened since 2016 unfortunately. I’ve added the details we researched, although the initial report was already pretty clear.